src/normalizer.ts

```diff
@@ -1,233 +1,246 @@
+}
+function isObjectType(schema: LinkedJSONSchema) {
+  return schema.properties !== undefined || hasType(schema, 'object') || hasType(schema, 'any')
+}
+function isArrayType(schema: LinkedJSONSchema) {
+  return (
+    schema.items !== undefined || schema.prefixItems !== undefined || hasType(schema, 'array') || hasType(schema, 'any')
+  )
+}
+
+rules.set('Remove `type=["null"]` if `enum=[null]`', schema => {
+  if (
+    Array.isArray(schema.enum) &&
+    schema.enum.some(e => e === null) &&
+    Array.isArray(schema.type) &&
+    schema.type.includes('null')
+  ) {
+    schema.type = schema.type.filter(type => type !== 'null')
+  }
+})
+
+rules.set('Destructure unary types', schema => {
+  if (schema.type && Array.isArray(schema.type) && schema.type.length === 1) {
+    schema.type = schema.type[0]
+  }
+})
+
+rules.set('Add empty `required` property if none is defined', schema => {
+  if (isObjectType(schema) && !('required' in schema)) {
+    schema.required = []
+  }
+})
+
+rules.set('Transform `required`=false to `required`=[]', schema => {
+  if (schema.required === false) {
+    schema.required = []
+  }
+})
+
+rules.set('Default additionalProperties', (schema, _, options) => {
+  if (isObjectType(schema) && !('additionalProperties' in schema) && schema.patternProperties === undefined) {
+    schema.additionalProperties = options.additionalProperties
+  }
+})
+
+rules.set('Transform id to $id', (schema, fileName) => {
+  if (!isSchemaLike(schema)) {
+    return
+  }
+  if (schema.id && schema.$id && schema.id !== schema.$id) {
+    throw ReferenceError(
+      `Schema must define either id or $id, not both. Given id=${schema.id}, $id=${schema.$id} in ${fileName}`,
+    )
+  }
+  if (schema.id) {
+    schema.$id = schema.id
+    delete schema.id
+  }
+})
+
+rules.set('Add an $id to anything that needs it', (schema, fileName, _options, _key, dereferencedPaths) => {
+  if (!isSchemaLike(schema)) {
+    return
+  }
+
+  // Top-level schema
+  if (!schema.$id && !schema[Parent]) {
+    schema.$id = toSafeString(justName(fileName))
+    return
+  }
+
+  // Sub-schemas with references
+  if (!isArrayType(schema) && !isObjectType(schema)) {
+    return
+  }
+
+  // We'll infer from $id and title downstream
+  // TODO: Normalize upstream
+  const dereferencedName = dereferencedPaths.get(schema)
+  if (!schema.$id && !schema.title && dereferencedName) {
+    schema.$id = toSafeString(justName(dereferencedName))
+  }
+
+  if (dereferencedName) {
+    dereferencedPaths.delete(schema)
+  }
+})
+
+rules.set('Escape closing JSDoc comment', schema => {
+  escapeBlockComment(schema)
+})
+
+rules.set('Rename `prefixItems` to `items` and `items` to `additionallItems`', (schema, _, options) => {
+  if (isArrayType(schema) && options.usePrefixItems) {
+    // @ts-expect-error this is a simple renaming procedure, there is in no need to change the `schema.items` type to allow for boolean values

```

Comment:
Please don't expect an error. Instead, make the types safer or use a type assertion (`a as A`).